### DISCUSSION TITLE
Why is the Windows App SDK's MSIX dynamic dependency API limited to non-packaged applications?

### DISCUSSION THREAD

**Discussion by Gaoyifei1011** [microsoft/WindowsAppSDK#4819] [General]
Why is the Windows App SDK's MSIX dynamic dependency API limited to non-packaged applications?

### -------------------------------------------

为什么 Windows App SDK 的 MSIX 动态依赖项 API 仅限于非打包应用？

-------------------------------------------------------------------------

My project uses WinUI 2 and WinUI 3 together, WinUI 2 and WinUI 3 contain both Microshot.ui.xaml.dll dependency libraries, and I need to use the MSIX dynamic dependency API to resolve this conflict. When I use the Windows App SDK dynamic dependency API, I find that the Windows App SDK dynamic dependency API is limited to non-packaged applications, and packaged applications will directly fail to call this function. So I would like to ask if this API can be opened to packaging applications in the future?

### -------------------------------------------

我的项目共同使用了 WinUI 2 和 WinUI 3，WinUI 2 和 WinUI 3 共同包含 Microsoft.UI.Xaml.dll 依赖库，需要使用 MSIX 动态依赖项 API 来解决这一冲突。当我使用 Windows App SDK 动态依赖项 API 时，发现 Windows App SDK 动态依赖项 API 仅限于非打包应用，打包应用调用该函数会直接失败。所以想问一下未来这个 API 可以开放给打包应用使用吗？

-------------------------------------------------------------------------

![Image](https://github.com/user-attachments/assets/d7a13268-8264-404b-854b-1d96e4f257df)

https://github.com/microsoft/WindowsAppSDK/blob/aa9c8b8730ea23b656e882766839e090335e2f93/dev/DynamicDependency/API/MsixDynamicDependency.cpp#L36

**Comment by RDMacLachlan** [comment]
Can you provide some context on what your scenario for this is?

**Reply by Gaoyifei1011** [reply]
> Can you provide some context on what your scenario for this is?

https://github.com/Gaoyifei1011/GetStoreApp

这是我的项目，我的项目是使用 MSIX 打包的应用程序。在这个 MSIX 打包应用程序中共包含多个可执行文件，其中一个是 WinUI 3（桌面应用程序），另外一个是辅助应用程序（UWP 应用程序）。这个辅助应用程序主要是帮助我完成固定应用程序到任务栏的一些操作（调用 TaskbarManager API，该 API 需要在 UWP CoreApplication 模式下正常使用）。
当我的 UWP 应用引入 Microsoft.UI.Xaml(WinUI 2) Nuget 包时，由于桌面应用程序使用的是 Windows App SDK (WinUI 3) Nuget 包。在编译应用程序完成后，并部署应用到本地设备测试运行时，有趣的事情发生了。由于 WinUI 2 和 Windows App SDK 共同包含 Microsoft.UI.Xaml.dll，而且 WinUI 2 包的包依赖在 Windows App SDK 包依赖的前面（AppxManifest.xml PackageDependency 显示的位置），导致 WinUI 3 桌面应用程序误加载 WinUI 2 的 Microsoft.UI.Xaml.dll 从而导致应用程序启动失败。这时需要使用 MSIX 动态依赖项强行修改 WinUI 3 桌面应用程序的包图依赖顺序并将将 Windows App SDK 包依赖排在 WinUI 2 的前面。
这就是我需要在 MSIX 打包应用程序中引入 MSIX 动态依赖项的原因。然而我尝试调用 Windows App SDK 的动态依赖项 API 时，却发生了异常。查看源代码时发现，该 API 调用仅限于非打包应用进程。所以希望 Windows App SDK 团队能解除这一限制，并允许打包应用程序进程使用这一 API。

##### ------------------------------------------

This is my project, and my project is an application packaged with MSIX. There are several executables in this MSIX packaged application, one of which is WinUI 3 (desktop application) and the other is a secondary application (UWP application). This helper app is mainly to help me pin the app to the taskbar (calling the TaskbarManager API, which needs to work properly in UWP CoreApplication mode).
When my UWP app introduces the Microsoft.UI.Xaml(WinUI 2) Nuget package, the desktop app is using the Windows App SDK (WinUI 3) Nuget package. After compiling the application and deploying the app to the local device test run, something interesting happened. Because WinUI 2 and the Windows App SDK both contain Microsoft.UI.Xaml.dll, and the WinUI 2 package dependency is in front of the Windows App SDK package dependency (where the AppxManifest.xml PackageDependency appears), the WinUI 3 desktop application mistakenly loads WinUI 2 Microsoft.UI.Xaml.dll causes the application to fail to start. In this case, you need to use MSIX dynamic dependencies to forcibly modify the package diagram dependency order of the WinUI 3 desktop application, and rank the Windows App SDK package dependencies before WinUI 2.
That's why I need to introduce MSIX dynamic dependencies in my MSIX packaged application. However, when I try to call the Windows App SDK's dynamic dependency API, I get an exception. When you look at the source code, you can see that the API call is limited to unpackaged app processes. So hopefully the Windows App SDK team will remove this restriction and allow packaged application processes to use this API.

----------------------------------------

![2afef7a8efa9b86fdf6b21d4d7f57cd3](https://github.com/user-attachments/assets/4edd5cec-f3b6-4111-b0ee-2f03d1dc0277)
GetStoreApp 是 WinUI 3 桌面应用程序，GetStoreAppWebView 是 UWP 应用程序（使用 WinUI 2），GetStoreAppWebView 应用充当 UWP 应用固定应用到任务栏这一行为。

##### ------------------------------------------

GetStoreApp is a WinUI 3 desktop app, GetStoreAppWebView is a UWP app (using WinUI 2), and the GetStoreAppWebView app acts as a UWP app pinning apps to taskbar behavior.

![{D223E34E-F07C-417A-B35E-6EA741246DB2}](https://github.com/user-attachments/assets/d13e851f-ca0c-4b70-9d47-56625f20a3bb)

----------------------------------------

API-related calls （API 相关调用）
![{9DB76368-F2DA-4B24-9080-FCCE33A7BAC8}](https://github.com/user-attachments/assets/0eaaf8d0-a0ab-4138-b868-4f313d4486ea)
![{9924A980-D39C-4E7A-8ECF-B6C2EF0E930B}](https://github.com/user-attachments/assets/99e72ad5-8719-430b-8bb1-3f6ace1bec2b)
https://github.com/Gaoyifei1011/GetStoreApp/blob/main/GetStoreApp/Program.cs#L73

----------------------------------------

Project Nuget package usage （项目 Nuget 包使用情况）
![{C87B4827-5921-41A3-8EE6-2F78025601DA}](https://github.com/user-attachments/assets/c6056e93-28fc-446b-9e42-d3034d35e48a)
